File: demo/app/src/main/java/com/timper/lonelysword/app/data/remote/ServiceFactor.java
//package com.nio.so.service.steward.data.repository;
//
//import com.facebook.stetho.okhttp3.StethoInterceptor;
//import com.google.gson.FieldNamingPolicy;
//import com.google.gson.Gson;
//import com.google.gson.GsonBuilder;
//import java.io.IOException;
//import java.util.concurrent.TimeUnit;
//import okhttp3.Interceptor;
//import okhttp3.OkHttpClient;
//import okhttp3.Request;
//import okhttp3.Response;
//import okhttp3.logging.HttpLoggingInterceptor;
//import retrofit2.Retrofit;
//import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
//import retrofit2.converter.gson.GsonConverterFactory;
//
///**
// * User: tangpeng.yang
// * Date: 20/03/2018
// * Description:
// * FIXME
// */
//public class ServiceFactor {
//
//  public static <T> T createService(String host, Boolean isDebug, Class<T> clazz) {
//    OkHttpClient okHttpClient = createOkHttpClient(isDebug, createLoggingInterceptor(isDebug));
//    return createService(host, okHttpClient, createGson(), clazz);
//  }
//
//  private static <T> T createService(String host, OkHttpClient okHttpClient, Gson gson, Class<T> clazz) {
//    Retrofit retrofit = new Retrofit.Builder().baseUrl(host)
//        .client(okHttpClient)
//        .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
//        .addConverterFactory(GsonConverterFactory.create(gson))
//        .build();
//    return retrofit.create(clazz);
//  }
//
//  private static OkHttpClient createOkHttpClient(Boolean isDebug, HttpLoggingInterceptor httpLoggingInterceptor) {
//    OkHttpClient.Builder builder = new OkHttpClient.Builder();
//    builder.addInterceptor(httpLoggingInterceptor).addInterceptor(createHeadInterceptor());
//    //if (isDebug) {
//    builder.addNetworkInterceptor(new StethoInterceptor());
//    //}
//    return builder.connectTimeout(120, TimeUnit.SECONDS).readTimeout(120, TimeUnit.SECONDS).build();
//  }
//
//  private static Gson createGson() {
//    return new GsonBuilder().setLenient()
//        .setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
//        .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
//        .create();
//  }
//
//  private static Interceptor createHeadInterceptor() {
//    return new Interceptor() {
//      @Override public Response intercept(Chain chain) throws IOException {
//        Request original = chain.request();
//        Request request = original.newBuilder()
//            .header("Content-Type", "application/json;charset=UTF-8")
//            .method(original.method(), original.body())
//            .build();
//        return chain.proceed(request);
//      }
//    };
//  }
//
//  private static HttpLoggingInterceptor createLoggingInterceptor(Boolean isDebug) {
//    HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
//    if (isDebug) {
//      logging.setLevel(HttpLoggingInterceptor.Level.BODY);
//    } else {
//      logging.setLevel(HttpLoggingInterceptor.Level.NONE);
//    }
//    return logging;
//  }
//}

File: demo/module/src/main/java/com/timper/module/data/remote/ServiceFactor.kt
package com.timper.module.data.remote

import com.facebook.stetho.okhttp3.StethoInterceptor
import com.google.gson.FieldNamingPolicy
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.jakewharton.retrofit2.converter.kotlinx.serialization.asConverterFactory
import kotlinx.serialization.json.Json
import okhttp3.Interceptor
import okhttp3.MediaType
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

/**
 * User: tangpeng.yang
 * Date: 20/03/2018
 * Description:
 * FIXME
 */
class ServiceFactor private constructor() {
    companion object {
        val instance = SingletonHolder.holder
    }

    private object SingletonHolder {
        val holder = ServiceFactor()
    }

    val retrofit: Retrofit by lazy {
        Retrofit.Builder()
            .baseUrl("https://www.wanandroid.com")
            .client(createOkHttpClient(true, createLoggingInterceptor(true)))
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
//            .addConverterFactory(Json.asConverterFactory(MediaType.get("application/json")))
            .addConverterFactory(GsonConverterFactory.create(createGson()))
            .build()
    }

    fun <T> createService(clazz: Class<T>): T {
        return retrofit.create(clazz)
    }

    private fun createOkHttpClient(isDebug: Boolean?, httpLoggingInterceptor: HttpLoggingInterceptor): OkHttpClient {
        val builder = OkHttpClient.Builder()
        builder.addInterceptor(httpLoggingInterceptor)
            .addInterceptor(createHeadInterceptor())
        isDebug?.let {
            if (it) {
                builder.addNetworkInterceptor(StethoInterceptor())
            }
        }
        builder.retryOnConnectionFailure(true)
        return builder.connectTimeout(120, TimeUnit.SECONDS)
            .readTimeout(120, TimeUnit.SECONDS)
            .build()
    }

    private fun createGson(): Gson {
        return GsonBuilder().setLenient()
            .setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
            .setFieldNamingPolicy(FieldNamingPolicy.IDENTITY)
            .create()
    }

    private fun createHeadInterceptor(): Interceptor {
        return Interceptor { chain ->
            val original = chain.request()
            val request = original.newBuilder()
                .header("Content-Type", "application/json;charset=UTF-8")
                .method(original.method(), original.body())
                .build()
            chain.proceed(request)
        }
    }

    private fun createLoggingInterceptor(isDebug: Boolean): HttpLoggingInterceptor {
        val logging = HttpLoggingInterceptor()
        if (isDebug) {
            logging.level = HttpLoggingInterceptor.Level.BODY
        } else {
            logging.level = HttpLoggingInterceptor.Level.NONE
        }
        return logging
    }
}
